### PR TITLE
add support for postgis geometry and geography fields and their data …

### PR DESCRIPTION
**Pull request by nikcub**
…types in a new geoType column

i'll be writing tests and adding them as soon as I update the framework so that it is more generic for other use cases - but we're using this in production atm.

### CODE REVIEW

**Comment by ozum** [comment]
Thank you for the contribution.

**Comment by ozum** [comment]
Hi,

After merge, I realized, that it throws exception if PostgreSQL is installed without PostGIS extension. I had to revert back.

Could you please make it compatible with PostgreSQL without PostGIS extension.

Thanks,

**Comment by jayarjo** [comment]
@ozum What was the exception? I didn't try it, but from what I see, it has all pre-cautions in place.

File: lib/column.js [hunk 1]
```diff
@@ -10,15 +10,15 @@ let parameters  = [
...
+    'precision', 'scale', 'type', 'userDefinedType', 'geoType'
 ];
 
 let attributes  = [
     'allowNull', 'arrayDimension', 'arrayType',
     'defaultWithTypeCast', ['description', 'comment'], ['descriptionData', 'commentData'], ['enumValues', 'enumLabels'],
     'domainName', 'domainSchema', ['isAutoIncrement', 'isSerial'],
     'length', 'name', 'onDelete', 'onUpdate', ['parent', 'table'],
-    'precision', 'scale', 'type', 'userDefinedType'
+    'precision', 'scale', 'type', 'userDefinedType', 'geoType'
```
Comment (jayarjo) [inline lib/column.js:21]: And where is this `geoType` populated?

**Comment by ozum** [comment]
I couldn't remember, since 4 years passed. All I can remember is, it fails if extension is not used in database. Also I don't have exp. in PostGIS.

**Comment by btiernay** [comment]
Any chance of bringing this back but making support opt-in?

**Comment by ozum** [comment]
> Any chance of bringing this back but making support opt-in?

Since I don't have PostGIS experience, I need a complete PR from a contributor.

**Comment by btiernay** [comment]
What about support for not erroring out when there is an unknown column type? Can that be made configurable perhaps?  Here's what I'm observing:
```
Error: 'geography' cannot be found in CompositeType's name.
    at Proxy.get (project/node_modules/indexable-array/dist/index.js:491:19)
    at new Column (project/node_modules/pg-structure/dist/pg-structure/column.js:38:128)
    at project/node_modules/pg-structure/dist/main.js:184:29
    at Array.forEach (<anonymous>)
    at addColumns (project/node_modules/pg-structure/dist/main.js:180:10)
    at addObjects (project/node_modules/pg-structure/dist/main.js:333:5)
    at Object.pgStructure [as default] (project/node_modules/pg-structure/dist/main.js:376:5)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    ```

**Comment by ozum** [comment]
> What about support for not erroring out when there is an unknown column type? Can that be made configurable perhaps?

It's not that simple. We have to handle other columns that refer (FK) to this and other inner details. It could cause lots of inconsistency.

**Comment by btiernay** [comment]
I understand the concern, but perhaps it could be configurable such that a more reasonable placeholder could be used. I was able to get around this for my use case where I'm using `pg-tables-to-jsonschema` which in turn uses `pg-structurizer`, but it's a big hack:
```javascript
// HACK: This is needed due to https://github.com/ozum/pg-structure/pull/4#issuecomment-1211291313
const oldGet = IndexableArray.prototype.get
IndexableArray.prototype.get = function (...args) {
  try {
    return oldGet.apply(this, args)
  } catch (e) {
    if (args.length === 1 && args[0] === 'geography') {
      console.error('Error processing', args, e instanceof Error ? e.message : e)
      return {
        schema: 'dummy',
        typeName: 'geography',
        length: undefined,
        precision: undefined,
        scale: undefined,
      }
    } else {
      throw e
    }
  }
}
```

Having a mechanism that can support this would be preferred.

**Comment by ozum** [comment]
> but perhaps it could be configurable such that a more reasonable placeholder could be used

Unfortunately, the "pg-structure" engine could not workaround lots of edge cases. There are unlimited numbers of different database designs including composite foreign keys, index referrals, etc...

**Pull request merged**